### PR TITLE
feat: accept Uint8Arrays input in place of Buffers

### PR DESCRIPTION
**Pull request by Gozala**
This pull request relaxes input type from Buffer to Uint8Array. It does however depends on https://github.com/multiformats/js-multibase/pull/61

### CODE REVIEW

**Comment by Gozala** [comment]
> I've the same questions as @rvagg (thanks for bringing it up) [multiformats/js-multibase#61 (comment)](https://github.com/multiformats/js-multibase/pull/61#issuecomment-662362628):

I'm not sure what changes with `webpack@5` but:
1. It is only in tests
2. It is only relevant for <=node@10 I think. So if we don't care testing in older versions of node, I can remove it.

> Could you please add a test that is using a Buffer as input, to make sure it still works?

Ok, I'll do that

> Am I correct that this change depends on multiformats/js-multibase#61?

Yep, as description said 🤓

**Comment by Gozala** [comment]
- [x] Have test run with both `Buffer` and `Uint8Array` inputs.
  > Could you please add a test that is using a Buffer as input, to make sure it still works?
- [x] Imported `web-encodings` library to deal with that
  > Does the require('util') mean we need to include it in the dependencies list to make sure we have it available for webpack@5?

**Comment by mikeal** [comment]
If it’s only in tests it’s probably fine, the primary concern is that bundlers are not injecting polyfills for Node.js anymore.

At some point we may even want it out of the tests, and we should all be getting in the habit of working without Node.js stdlib as much as possible, but it’s not a blocker.

We’ve cut Node.js v10 elsewhere but I’m not 100% sure what our dependencies have done that rely on this library.

**Comment by Gozala** [comment]
> If it’s only in tests it’s probably fine, the primary concern is that bundlers are not injecting polyfills for Node.js anymore.
> 
> At some point we may even want it out of the tests, and we should all be getting in the habit of working without Node.js stdlib as much as possible, but it’s not a blocker.
> 
> We’ve cut Node.js v10 elsewhere but I’m not 100% sure what our dependencies have done that rely on this library.

[web-encoding](https://github.com/gozala/web-encoding) should take care of providing `TextEncoder` / `TextDecoder` across the board. And it's tested on node v10 as well as 12 and 14

File: src/index.js [hunk 1]
```diff
@@ -22,15 +25,19 @@ exports.codes = Object.freeze(codes)
...
  *
- * @param {Buffer} hash
+ * @param {Uint8Array} hash
  * @returns {string}
  */
 exports.toHexString = function toHexString (hash) {
-  if (!Buffer.isBuffer(hash)) {
-    throw new Error('must be passed a buffer')
+  if (!(hash instanceof Uint8Array)) {
+    throw new Error('must be passed an Uint8Array')
```
Comment (vmx) [inline src/index.js:33]: Nitpick (not critical, more fyi). It's "*a* Uint8array" as it depends on the pronunciation, "Uint" starts with "ju…".

File: src/index.js [hunk 2]
```diff
@@ -22,15 +25,19 @@ exports.codes = Object.freeze(codes)
 /**
  * Convert the given multihash to a hex encoded string.
  *
- * @param {Buffer} hash
+ * @param {Uint8Array} hash
```
Comment (vmx) [inline src/index.js:28]: `Buffer` is also a `Uint8Array`, but should this perhaps be `Uint8Array|Buffer` for more clarity? I'm not sure about this, I leave this up to you, as you've spend more time on this kind of conversion also in other libraries.

File: src/util.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+'use strict'
```
Comment (vmx) [inline src/util.js:1]: It looks like you're not using this file anymore.

File: test/index.spec.js [hunk 4]
```diff
@@ -12,6 +12,9 @@ const mh = require('../src')
 const constants = require('../src/constants')
 const validCases = require('./fixtures/valid')
 const invalidCases = require('./fixtures/invalid')
+const textEncoder = typeof TextEncoder !== 'undefined'
```
Comment (vmx) [inline test/index.spec.js:15]: The test should also use `web-encoding`.

File: test/index.spec.js [hunk 5]
```diff
@@ -24,12 +27,27 @@ function sample (code, size, hex) {
   return Buffer.from(`${toHex(code)}${toHex(size)}${hex}`, 'hex')
 }
 
+const they = (description, test) => {
```
Comment (vmx) [inline test/index.spec.js:30]: That's a nice and smart way to test both.

File: src/index.js (hunk 1 above)
Comment (Gozala) [reply src/index.js:33]: I always thought if it was based on first latter being a vowel 🤯

File: test/index.spec.js [hunk 6]
```diff
@@ -78,15 +95,15 @@ describe('multihash', () => {
       expect(
         () => mh.toB58String('hello world')
       ).to.throw(
-        /must be passed a buffer/
+        /must be passed an Uint8Array/
```
Comment (vmx) [inline test/index.spec.js:98]: All occurrences of "an Uint8Array" needs to be changes, else tests fail :)

**Comment by Gozala** [comment]
> Approved, though I'd like CI to run and be green before it is merged.

Once #89 lands, I can run CI tests.

**Comment by achingbrain** [comment]
> Once #89 lands, I can run CI tests.

This has landed now and the tests are all green, but for the future you could have used that branch as the merge base for this PR which would have let you run CI before it was merged.

**Pull request merged**